src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayer.java

```diff
@@ -221,6 +224,88 @@ public void setTotalExperience(int exp) {
         this.experienceTotal = exp;
     }
 
+    /**
+     * {@link EntityPlayer#addExperienceLevel(int)} and
+     * {@link EntityPlayer#removeExperienceLevel(int)} don't update the total
+     * experience. This recalculates it for plugins to properly make use of it.
+     */
+    private void recalculateTotalExperience() {
+        if (!this.dontRecalculateExperience) {
+            int newExperienceInLevel = (int) (this.experience * this.xpBarCap());
+            this.experienceTotal = this.xpAtLevel(this.experienceLevel) + newExperienceInLevel;
+            this.experience = (float) newExperienceInLevel / this.xpBarCap();
+        }
+    }
+
+    @Inject(method = "addExperienceLevel", at = @At("RETURN"))
+    private void onAddExperienceLevels(int levels, CallbackInfo ci) {
+        recalculateTotalExperience();
+    }
+
+    @Inject(method = "removeExperienceLevel", at = @At("RETURN"))
+    private void onRemoveExperienceLevels(int levels, CallbackInfo ci) {
+        recalculateTotalExperience();
+    }
+
+    /**
+     * @author JBYoshi
+     * @reason This makes the experience updating more accurate and disables
+     * the totalExperience recalculation above for this method, which would
+     * otherwise have weird intermediate states.
+     */
+    @Overwrite
+    public void addExperience(int amount)
+    {
```

Comment:
No need to keep MCP formatting, shorten the line length by following sponge code style.